README.md
# SpaceEye Icons

Icons for the [SpaceEye application](https://github.com/KYDronePilot/SpaceEye)

## Info

This project stores source SVGs and build tools to convert them into the various
icons needed in the [SpaceEye app](https://github.com/KYDronePilot/SpaceEye).
Builds must be run on macOS machines, since the `iconutil` command (exclusive to
macOS) is needed to create `.icns` files.

## Related

- [SpaceEye](https://github.com/KYDronePilot/SpaceEye) - The main SpaceEye app
- [SpaceEyeSatelliteConfig](https://github.com/KYDronePilot/SpaceEyeSatelliteConfig) -
  Config of available satellite views for SpaceEye app
- [space-eye-mac-node-api](https://github.com/KYDronePilot/space-eye-mac-node-api) -
  Native Mac API for SpaceEye app, to set multi-monitor wallpaper images
- [space-eye-windows-node-api](https://github.com/KYDronePilot/space-eye-windows-node-api) -
  Native Windows API for SpaceEye app, to set multi-monitor wallpaper images


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic
Versioning](https://semver.org/spec/v2.0.0.html).

## [0.4.0] - 2021-03-28

### Added

- Windows taskbar light mode icon.

## [0.3.1] - 2020-12-29

### Fixed

- Mac toolbar icons weren't being treated as template images, causing them to
  look invisible when in light mode.

## [0.3.0] - 2020-12-26

### Added

- More info in README and more metadata in package.json
- Added this CHANGELOG

### Changed

- Reformatted files
- Publish to npmjs.com instead of GitHub Packages to avoid needing API keys when
  installing
- Cleaned up the project

## [0.2.3] - 2020-12-13

### Changed

- Less blurring on small icons

## [0.2.2] - 2020-12-13

### Changed

- Removed macOS-like padding from Windows icons

## [0.2.1] - 2020-12-12

### Fixed

- Bug where yarn was used to install packages instead of npm

## [0.2.0] - 2020-12-12

### Changed

- Improved look of app icon
  - Looks more like a macOS icon with a dark gradient border
- Do Apple-like icon padding at build time, rather than having it hard-coded in
  the SVG
- Majorly refactored Gulp code
- Blurred small icons so they look less blocky

## [0.1.4] - 2020-11-16

### Added

- Build appx icons

### Changed

- Use Typescript for Gulp

## [0.1.3] - 2020-11-08

### Added

- Icon output for "About this app" page

### Changed

- Adjusted app icon so it looks better on Macs

## [0.1.1] - 2020-11-08

### Changed

- Removed app icon background and made it bigger

## [0.1.0] - 2020-11-08

### Changed

- Better icon names

## [0.0.1] - 2020-11-08

Initial release


gulpfile.js
const { exec } = require("child_process");
const { promisify } = require("util");
const { parallel, series } = require("gulp");
const fs = require("fs");
const path = require("path");

const asyncExec = promisify(exec);
const asyncCopyFile = promisify(fs.copyFile);

async function buildStandardToolbarIcon() {
  await asyncExec(
    "inkscape -w 20 -h 20 src/toolbar.svg --export-filename dist/mac_toolbar.png"
  );
}

async function buildRetinaToolbarIcon() {
  await asyncExec(
    "inkscape -w 40 -h 40 src/toolbar.svg --export-filename dist/mac_toolbar@2x.png"
  );
}

async function clean() {
  if (fs.existsSync("./dist")) {
    fs.rmdirSync("./dist", { recursive: true });
  }
  if (fs.existsSync("./build")) {
    fs.rmdirSync("./build", { recursive: true });
  }
  await Promise.all([asyncExec("mkdir ./dist"), asyncExec("mkdir ./build")]);
}

async function svgToPng(width, height, src, dest) {
  await asyncExec(
    `inkscape -w ${width} -h ${height} ${src} --export-filename ${dest}`
  );
}

async function svgToWindowsIco(src, dest) {
  const name = path.parse(dest).name;
  const sizes = [16, 24, 32, 48, 64, 72, 96, 128, 180, 256];
  await Promise.all(
    sizes.map((size) => svgToPng(size, size, src, `build/${name}${size}.png`))
  );
  await asyncExec(
    `convert ${sizes
      .map((size) => `build/${name}${size}.png`)
      .join(" ")} ${dest}`
  );
}

async function svgToMacIcns(src, dest) {
  const name = path.parse(dest).name;
  const sizes = [16, 32, 64, 128, 256, 512, 1024];
  const sizeNames = {
    16: ["16x16"],
    32: ["16x16@2x", "32x32"],
    64: ["32x32@2x"],
    128: ["128x128"],
    256: ["128x128@2x", "256x256"],
    512: ["256x256@2x", "512x512"],
    1024: ["512x512@2x"],
  };
  const iconsetDir = `build/${name}.iconset`;
  fs.mkdirSync(iconsetDir);
  await Promise.all(
    sizes.map((size) => svgToPng(size, size, src, `build/${name}${size}.png`))
  );
  await Promise.all(
    sizes.map(async (size) =>
      Promise.all(
        sizeNames[size].map(async (sizeName) =>
          asyncCopyFile(
            `build/${name}${size}.png`,
            path.join(iconsetDir, `icon_${sizeName}.png`)
          )
        )
      )
    )
  );
  await asyncExec(`iconutil -c icns ${iconsetDir} -o ${dest}`);
}

async function buildWindowToolbarIco() {
  await svgToWindowsIco("src/toolbar.svg", "dist/windows_toolbar.ico");
}

async function buildWindowAppIco() {
  await svgToWindowsIco("src/app.svg", "dist/windows_app.ico");
}

async function buildMacAppIcns() {
  await svgToMacIcns("src/app.svg", "dist/mac_app.icns");
}

async function buildInfoIcon() {
  await asyncExec(
    "inkscape -w 256 -h 256 src/app.svg --export-filename dist/info_app.png"
  );
}

exports.build = series(
  clean,
  parallel(
    buildStandardToolbarIcon,
    buildRetinaToolbarIcon,
    buildWindowToolbarIco,
    buildWindowAppIco,
    buildMacAppIcns,
    buildInfoIcon
  )
);


gulpfile.ts
import { exec } from 'child_process'
import { promisify } from 'util'
import { parallel, series } from 'gulp'
import fs from 'fs'
import fse from 'fs-extra'
import path from 'path'

const asyncExec = promisify(exec)
const asyncCopyFile = promisify(fs.copyFile)

const SRC = path.join(__dirname, 'src')
const BUILD = path.join(__dirname, 'build')
const DIST = path.join(__dirname, 'dist')

const APP_SVG = path.join(SRC, 'app.svg')
const TOOLBAR_SVG = path.join(SRC, 'toolbar.svg')
const APPX = path.join(DIST, 'appx')

/**
 * Clean build/dist dirs, preparing for new build.
 */
async function clean() {
    await Promise.all([fse.emptyDir(BUILD), fse.emptyDir(DIST)])
    await fse.ensureDir(APPX)
}

/**
 * Add padding to a PNG (in-place).
 *
 * @param paddingX - Padding in X dimension on each side
 * @param paddingY - Padding in Y dimension on each side
 * @param pngPath - Path to the PNG
 */
async function padPng(paddingX: number, paddingY: number, pngPath: string) {
    await asyncExec(
        `magick mogrify -bordercolor transparent -border ${paddingX}x${paddingY} -format png ${pngPath}`
    )
}

/**
 * Convert an SVG to a PNG.
 *
 * - Assumes SVGs are squares
 * - If dimensions are not square, padding is added for extra space
 *
 * @param width - Width of the PNG
 * @param height - Height of the PNG
 * @param svgPath - Path to the SVG
 * @param pngPath - Path to the PNG
 */
async function buildSvgToPng(
    width: number,
    height: number,
    svgPath: string,
    pngPath: string,
    preScalingFactor: number = 1,
    blurSigma: number | undefined = undefined,
    invert: boolean = false
) {
    // Build to PNG square of minDimension x minDimension
    const minDimension = Math.min(width, height)
    // Scale for initial dimensions (if using blurring)
    const preScaledDimensions = minDimension * preScalingFactor
    const initialDimensions = preScaledDimensions.toString()
    await asyncExec(
        `inkscape -w ${initialDimensions} -h ${initialDimensions} "${svgPath}" --export-filename "${pngPath}"`
    )
    // Invert color if necessary
    if (invert) {
        await asyncExec(`magick mogrify -channel RGB -negate ${pngPath}`)
    }
    if (blurSigma !== undefined) {
        // Blur the image
        await asyncExec(`magick mogrify -blur 0x${blurSigma.toFixed(1)} ${pngPath}`)
        // Resize down if scaling done
        if (minDimension !== preScaledDimensions) {
            await asyncExec(`magick mogrify -resize ${minDimension}x${minDimension} ${pngPath}`)
        }
    }
    // If not the same dimensions, add padding to take up the space
    if (width !== height) {
        await padPng(
            Math.round((width - minDimension) / 2),
            Math.round((height - minDimension) / 2),
            pngPath
        )
    }
}

/**
 * Build a PNG file with standard Mac icon padding.
 *
 * @param width - Width of the PNG
 * @param height - Height of the PNG
 * @param svgPath - Path to the SVG
 * @param pngPath - Path to the PNG
 */
async function buildPngWithMacIconPadding(
    width: number,
    height: number,
    svgPath: string,
    pngPath: string,
    preScalingFactor: number = 1,
    blurSigma: number | undefined = undefined
) {
    const totalPadding = Math.round((33 * height) / 256)
    const paddingAmount = totalPadding / 2
    await buildSvgToPng(
        width - totalPadding,
        height - totalPadding,
        svgPath,
        pngPath,
        preScalingFactor,
        blurSigma
    )
    await padPng(paddingAmount, paddingAmount, pngPath)
}

/**
 * Build toolbar icons for Mac.
 */
async function buildMacToolbarIcons() {
    await Promise.all([
        buildSvgToPng(20, 20, TOOLBAR_SVG, path.join(DIST, 'MacToolbarTemplate.png'), 16, 8),
        buildSvgToPng(40, 40, TOOLBAR_SVG, path.join(DIST, 'MacToolbarTemplate@2x.png'), 8, 6),
    ])
}

/**
 * Build a SVG to a Windows ICO file, with all appropriate sizes.
 *
 * @param svgPath - Path to SVG
 * @param icoPath - Path to ICO file
 */
async function buildSvgToWindowsIco(svgPath: string, icoPath: string, invert: boolean = false) {
    // Create a dir to save the initial PNG files in
    const name = path.parse(icoPath).name
    const buildAssets = path.join(BUILD, name)
    fse.ensureDir(buildAssets)
    const sizes: [number, number, number | undefined][] = [
        [16, 16, 6],
        [24, 16, 5],
        [32, 10, 4],
        [48, 8, 2],
        [64, 5, 2],
        [72, 1, undefined],
        [96, 1, undefined],
        [128, 1, undefined],
        [180, 1, undefined],
        [256, 1, undefined],
    ]
    // Build each size to a PNG
    await Promise.all(
        sizes.map(([size, initialScale, blurSigma]) =>
            buildSvgToPng(
                size,
                size,
                svgPath,
                path.join(buildAssets, `${size}.png`),
                initialScale,
                blurSigma,
                invert
            )
        )
    )
    // Convert to ICO
    await asyncExec(
        `convert ${sizes
            .map(([size, _, __]) => '"' + path.join(buildAssets, `${size}.png`) + '"')
            .join(' ')} "${icoPath}"`
    )
}

/**
 * Build a SVG to a Mac ICNS file, with appropriate sizes.
 *
 * @param svgPath - Path to SVG
 * @param icnsPath - Path to ICNS file
 */
async function svgToMacIcns(svgPath: string, icnsPath: string) {
    // Create a dir to save the initial PNG files in
    const name = path.parse(icnsPath).name
    const buildAssets = path.join(BUILD, name)
    fse.ensureDir(buildAssets)
    const sizes: [number, number, number | undefined][] = [
        [16, 16, 6],
        [32, 10, 4],
        [64, 5, 2],
        [128, 1, undefined],
        [256, 1, undefined],
        [512, 1, undefined],
        [1024, 1, undefined],
    ]
    // Names for each size
    const sizeNames: { [key: number]: string[] } = {
        16: ['16x16'],
        32: ['16x16@2x', '32x32'],
        64: ['32x32@2x'],
        128: ['128x128'],
        256: ['128x128@2x', '256x256'],
        512: ['256x256@2x', '512x512'],
        1024: ['512x512@2x'],
    }
    // Dir to save the renamed files to
    const iconsetDir = path.join(buildAssets, 'images.iconset')
    fse.ensureDir(iconsetDir)
    // Build initial PNGs
    await Promise.all(
        sizes.map(([size, initialScale, blurSigma]) =>
            buildPngWithMacIconPadding(
                size,
                size,
                svgPath,
                path.join(buildAssets, `${size}.png`),
                initialScale,
                blurSigma
            )
        )
    )
    // Copy to proper file names
    await Promise.all(
        sizes.map(async ([size, _, __]) =>
            Promise.all(
                sizeNames[size].map(async (sizeName) =>
                    asyncCopyFile(
                        path.join(buildAssets, `${size}.png`),
                        path.join(iconsetDir, `icon_${sizeName}.png`)
                    )
                )
            )
        )
    )
    // Finally, build to an icns file
    await asyncExec(`iconutil -c icns "${iconsetDir}" -o "${icnsPath}"`)
}

/**
 * Build toolbar ICO for Windows.
 */
async function buildWindowsToolbarIco() {
    await buildSvgToWindowsIco(TOOLBAR_SVG, path.join(DIST, 'windows_toolbar.ico'))
}

/**
 * Build light toolbar ICO for Windows.
 */
async function buildWindowsLightToolbarIco() {
    await buildSvgToWindowsIco(TOOLBAR_SVG, path.join(DIST, 'windows_toolbar_light.ico'), true)
}

/**
 * Build app ICO for Windows.
 */
async function buildWindowsAppIco() {
    await buildSvgToWindowsIco(APP_SVG, path.join(DIST, 'windows_app.ico'))
}

/**
 * Build app ICNS for Mac.
 */
async function buildMacAppIcns() {
    await svgToMacIcns(APP_SVG, path.join(DIST, 'mac_app.icns'))
}

/**
 * Build info icon (for "About This App" page).
 */
async function buildInfoIcon() {
    await buildSvgToPng(256, 256, APP_SVG, path.join(DIST, 'info_app.png'))
}

/**
 * Build app icons for APPX (Microsoft Store) builds.
 */
async function buildUwpIcons() {
    return Promise.all([
        buildSvgToPng(512, 512, APP_SVG, path.join(APPX, 'BadgeLogo.png')),
        buildSvgToPng(310, 310, APP_SVG, path.join(APPX, 'LargeTile.png')),
        buildSvgToPng(71, 71, APP_SVG, path.join(APPX, 'SmallTile.png'), 4, 2),
        buildSvgToPng(44, 44, APP_SVG, path.join(APPX, 'Square44x44Logo.png'), 8, 4),
        buildSvgToPng(150, 150, APP_SVG, path.join(APPX, 'Square150x150Logo.png')),
        buildSvgToPng(64, 64, APP_SVG, path.join(APPX, 'StoreLogo.png'), 4, 2),
        buildSvgToPng(310, 150, APP_SVG, path.join(APPX, 'Wide310x150Logo.png')),
    ])
}

export const build = series(
    clean,
    parallel(
        buildMacToolbarIcons,
        buildWindowsToolbarIco,
        buildWindowsLightToolbarIco,
        buildWindowsAppIco,
        buildMacAppIcns,
        buildInfoIcon,
        buildUwpIcons
    )
)
